### PR TITLE
Update account_data when creating a new dm room

### PR DESCRIPTION
**Pull request by alfogrillo**
This PR put some logic in the Client type to update the `m.direct` account_data when creating a new dm room.

Fixes https://github.com/matrix-org/matrix-rust-sdk/issues/1674

### CODE REVIEW

File: crates/matrix-sdk/src/client/mod.rs [hunk 1]
```diff
@@ -1691,12 +1691,24 @@ impl Client {
     /// assert!(client.create_room(request).await.is_ok());
     /// # });
     /// ```
-    pub async fn create_room(&self, request: create_room::v3::Request) -> HttpResult<room::Joined> {
+    pub async fn create_room(&self, request: create_room::v3::Request) -> Result<room::Joined> {
```
Comment (alfogrillo) [inline crates/matrix-sdk/src/client/mod.rs:1694]: The idea here was to have a single method (bindings side) for both dm and normal rooms and let Rust to decide how to build the final ruma request.
If we think this approach isn't ideal I can expose a second method `create_dm_room` to UniFFI.
What do you think?

File: crates/matrix-sdk/src/client/mod.rs [hunk 2]
```diff
@@ -1691,12 +1691,24 @@ impl Client {
...
         let response = self.send(request, None).await?;
 
         let base_room =
             self.base_client().get_or_create_room(&response.room_id, RoomState::Joined).await;
-        Ok(room::Joined::new(self, base_room).unwrap())
+
+        let joined_room = room::Joined::new(self, base_room).unwrap();
+
+        // supporting just the DM chat with another person (no group dm chat)
+        if is_direct_room && invite.len() == 1 {
```
[the remark this answers is not in the document]
Comment (alfogrillo) [reply crates/matrix-sdk/src/client/mod.rs:1705]: Because I'm not sure how we are supposed to handle group chat.
If the user **A** creates a dm chat with **B** and **C** how would you updated **A**'s account data?

One option is like this:
```
{
  "content": {
    "B": [
      "room-id"
    ],
    "C": [
      "room-id"
    ]
  },
  "type": "m.direct"
}
```

But this also means that the A's client will think there is a dm chat **just** with B (or C).
This means that A's client cannot create a direct chat just with B (or C) anymore because we enforce just one DM chat for each pair of users.

[the remark this answers is not in the document]
Comment (alfogrillo) [reply crates/matrix-sdk/src/client/mod.rs:1705]: > Even if all Element clients didn't allow you add group DMs to m.direct, there's still tons of existing account data like this, and other clients will still be able to put group DMs in there as well.

I know, this doesn't work when other clients do different things, but at least we try hold it on Element.
What's the alternative for group chats anyway? Adding the room id just for the first member? Adding it for each member (like in the snippet in my previous comment)?

> For identifying an existing 1:1 room, you will have to iterate through known DM rooms with a user and look for one based on some heuristics (until we have canonical DMs).

Yeah possibly we want to rely on [this](https://github.com/matrix-org/matrix-rust-sdk/issues/1673) plus some check on the number of members (==2) maybe?

File: crates/matrix-sdk/src/encryption/mod.rs [hunk 3]
```diff
@@ -270,6 +268,24 @@ impl Client {
...
+        Ok(room)
+    }
+
+    /// Update the m.direct account data event.
+    ///
+    /// # Arguments
+    ///
+    /// * `room_id` - The room id of the DM room.
+    /// * `user_id` - The user id of the invitee for the DM room.
+    pub(crate) async fn update_m_direct_account_data(
```
[the remark this answers is not in the document]
Comment (alfogrillo) [reply crates/matrix-sdk/src/encryption/mod.rs]: Applied -> https://github.com/matrix-org/matrix-rust-sdk/pull/1665/commits/005464540f998a1e865b442ef01180f86e7da95e

File: crates/matrix-sdk/src/account.rs [hunk 4]
```diff
@@ -747,6 +747,40 @@ impl Account {
 
         Ok(self.client.send(request, None).await?)
     }
+
+    /// Marks the given room with `room_id` as "direct chat" with the given `user_id`.
+    /// This is done adding new the `room_id` to the list of DM chats with the user having id `user_id`.
```
Comment (jplatte) [inline crates/matrix-sdk/src/account.rs]:
```suggestion
    /// Marks the given room with `room_id` as "direct chat" with the given `user_id`.
    ///
    /// This is done by adding the new `room_id` to the list of DM chats with the user having id `user_id`.
```

**Comment by alfogrillo** [comment]
@jplatte I also need some help [here](https://github.com/matrix-org/matrix-rust-sdk/actions/runs/4438444146/jobs/7789780478?pr=1665) 🤔

File: crates/matrix-sdk/src/client/mod.rs [hunk 5]
```diff
@@ -1691,12 +1691,24 @@ impl Client {
...
+            if let Some(user) = invite.first() {
+                self.account().mark_as_dm(joined_room.room_id(), user).await?;
```
Comment (alfogrillo) [inline crates/matrix-sdk/src/client/mod.rs:1707]: Actually I'm not sure `await?` is the right thing here.
Shouldn't we return the room even if the account_data fails to be updated?

**Comment by jplatte** [comment]
> @jplatte I also need some help [here](https://github.com/matrix-org/matrix-rust-sdk/actions/runs/4438444146/jobs/7789780478?pr=1665) 🤔

This is a known compiler bug, will be fixed with the next nightly release (i.e. safe to ignore).

File: crates/matrix-sdk/src/client/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jplatte) [reply crates/matrix-sdk/src/client/mod.rs:1705]: > Yeah possibly we want to rely on [this](https://github.com/matrix-org/matrix-rust-sdk/issues/1673) plus some check on the number of members (==2) maybe?

I don't really know but that's not part of this PR anyways, right?

File: crates/matrix-sdk/src/client/mod.rs (hunk 5 above)
Comment (jplatte) [reply crates/matrix-sdk/src/client/mod.rs:1707]: Yes, ideally we would want to retry in the background if it fails, or something like that. You could do

```rust
if let Err(e) = self.account().mark_as_dm(...).await {
    // FIXME: Retry in the background
    error!("Failed to mark room as DM: {e}";
}
```

for now (using `tracing::error`).

File: crates/matrix-sdk/src/client/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (alfogrillo) [reply crates/matrix-sdk/src/client/mod.rs:1705]: Right

File: crates/matrix-sdk/src/client/mod.rs (hunk 5 above)
Comment (alfogrillo) [reply crates/matrix-sdk/src/client/mod.rs]: Applied: https://github.com/matrix-org/matrix-rust-sdk/pull/1665/commits/7cd89180ba61b1a4553370e9138fa171a8f01afb

File: crates/matrix-sdk/src/client/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (alfogrillo) [reply crates/matrix-sdk/src/client/mod.rs]: As discussed privately: in group DMs we add the room N times in m.direct (once for each invited user).
https://github.com/matrix-org/matrix-rust-sdk/commit/5c0b2b53683cb249c55ee48dc740c363d20b70ef

File: crates/matrix-sdk-crypto/src/identities/device.rs [hunk 6]
```diff
@@ -928,6 +928,7 @@ pub(crate) mod tests {
     }
 
     #[test]
+    #[allow(clippy::redundant_clone)]
```
Comment (alfogrillo) [inline crates/matrix-sdk-crypto/src/identities/device.rs:931]: @jplatte added this line to make Clippy happy even if it seems unrelated with this PR

**Review by jplatte — APPROVED** [review]
